src/selectors/TraversalSelector.h

```diff
@@ -0,0 +1,158 @@
+   * @return Smartpointer to the optimal traversal.
+   */
+  template <class PairwiseFunctor, bool useSoA, bool useNewton3>
+  std::unique_ptr<CellPairTraversal<ParticleCell>> getOptimalTraversal(PairwiseFunctor &pairwiseFunctor);
+
+  /**
+   * Evaluates to optimal traversal based on a given cell functor.
+   * @tparam PairwiseFunctor The functor that defines the interaction of two particles.
+   * @tparam useSoA
+   * @tparam useNewton3
+   * @param pairwiseFunctor The functor that defines the interaction of two particles.
+   */
+  template <class PairwiseFunctor, bool useSoA, bool useNewton3>
+  void tune(PairwiseFunctor &pairwiseFunctor);
+
+ private:
+  template <class PairwiseFunctor, bool useSoA, bool useNewton3>
+  std::vector<std::unique_ptr<CellPairTraversalInterface>> generateTraversals(PairwiseFunctor &pairwiseFunctor);
+
+  template <class PairwiseFunctor, bool useSoA, bool useNewton3>
+  std::unique_ptr<CellPairTraversal<ParticleCell>> chooseOptimalTraversal(
+      std::vector<std::unique_ptr<CellPairTraversalInterface>> &traversals);
+
+  // for each encountered cell processor save the optimal traversal. The cell processor is saved through its hash
+  std::unordered_map<size_t, TraversalOptions> _optimalTraversalOptions;
+  const std::array<unsigned long, 3> _dims;
+  const std::vector<TraversalOptions> _allowedTraversalOptions;
+};
+
+template <class ParticleCell>
+template <class PairwiseFunctor, bool useSoA, bool useNewton3>
+std::vector<std::unique_ptr<CellPairTraversalInterface>> TraversalSelector<ParticleCell>::generateTraversals(
+    PairwiseFunctor &pairwiseFunctor) {
+  std::vector<std::unique_ptr<CellPairTraversalInterface>> traversals;
+
+  for (auto &option : _allowedTraversalOptions) {
+    switch (option) {
+      case TraversalOptions::c08: {
+        traversals.push_back(
+            std::make_unique<C08Traversal<ParticleCell, PairwiseFunctor, useSoA, useNewton3>>(_dims, &pairwiseFunctor));
+        break;
+      }
+      case TraversalOptions::sliced: {
+        traversals.push_back(std::make_unique<SlicedTraversal<ParticleCell, PairwiseFunctor, useSoA, useNewton3>>(
+            _dims, &pairwiseFunctor));
+        break;
+      }
+      default: { AutoPasLogger->warn("Traversal type {} is not a known type!", option); }
+    }
+  }
+
+  assert(not traversals.empty());
+  return traversals;
+}
+
+template <class ParticleCell>
+template <class PairwiseFunctor, bool useSoA, bool useNewton3>
+std::unique_ptr<CellPairTraversal<ParticleCell>> TraversalSelector<ParticleCell>::chooseOptimalTraversal(
+    std::vector<std::unique_ptr<CellPairTraversalInterface>> &traversals) {
+  // remove traversals which are not applicable to the current domain size
+  traversals.erase(
+      std::remove_if(traversals.begin(), traversals.end(), [](auto const &t) { return not t->isApplicable(); }),
+      traversals.end());
+  assert(not traversals.empty());
+
+  // TODO: Autotuning goes here
+  // Tedious downcast
```

Comment:
why do you need this downcast?

Reply:
`getOptimalTraversal()` needs to return a pointer to a `CellPairTraversal` so `iteratePairwise` can be called on it.